### DISCUSSION TITLE
kamal custom builder for docker

### DISCUSSION THREAD

**Discussion by alexohre** [basecamp/kamal#914] [General]
hello folks, 
the title might not really convey the full information I wanted but here is the drill

so I have been playing with the kamal deployment tool recently and found it very interesting but at thesame time the build takes much time and requires more memory from my machine to process it. then I stumbled upon docker build cloud which build way faster than the local build.

here is the stats with my cloud build
initial build was 1ms 21s
the second build was just 28s

so now I was thinking if there is a way I can tell kamal to use a specific builder in my local machine that is connected to my build cloud instead of creating a new build all the time I run` kamal setup`

something like `Kamal setup --builder custom-build-name` to make the deploy process faster

and if its not an option now I think someone could make it possible soon

and finally a big thanks to the Rails community for all the good work so far.

**Comment by nickhammond** [comment]
Nice find! Kamal currently supports remote builders(https://kamal-deploy.org/docs/configuration/builders/#remote-configuration) but the builder name is kamal specific. A new option to the builder configuration would probably be needed or you might be able to create a context that points at Docker cloud and Kamal would just pick it up. Here are some builder examples as well https://kamal-deploy.org/docs/configuration/builder-examples/.

I'm putting the full build step here for documentation to help with a new configuration option PR.

Here's a full Docker cloud setup and build:
```
 docker login
 docker buildx create --driver cloud ORG/BUILDER_NAME
 docker buildx build --builder cloud-ORG-BUILDER_NAME --tag IMAGE .
```

Kamal creates a build context using a method based on the builder called `builder_name`
https://github.com/basecamp/kamal/blob/main/lib/kamal/commands/builder/native/remote.rb#L39
```
    def builder_name
      "kamal-#{config.service}-native-remote"
    end
```

So most likely a new driver(cloud?) that creates the build context with the cloud driver, creates a build context with that driver, and then builds with the cloud context.

**Reply by alexohre** [reply]
wow, thanks for the detailed response 

at last, someone finally understood me

```
 docker login
 docker buildx create --driver cloud ORG/BUILDER_NAME
 docker buildx build --builder cloud-ORG-BUILDER_NAME --tag IMAGE .
```

I have done the above already, it is just for me to tell Kamal to use my cloud-orgname builder to make life easy for my poor machine

**Reply by nickhammond** [reply]
@alexohre Sorry, that's not a working example it's just the steps that would need to happen in a PR.

**Comment by alexohre** [comment]
that would be awesome if it would be on the next update of kamal to speed
things a bit faster

**Comment by nickhammond** [comment]
It looks like this almost works, but needs a few small tweaks to get it going.

```
builder:
  arch: amd64
  driver: cloud nickhammond/hotdonuts
```

```
  INFO [551c2126] Running docker buildx inspect kamal-local-cloud nickhammond/hotdonuts as n@localhost
  ERROR (SSHKit::Command::Failed): docker exit status: 256
docker stdout: Nothing written
docker stderr: ERROR: "docker buildx inspect" requires at most 1 argument.
See 'docker buildx inspect --help'.

Usage:  docker buildx inspect [NAME]

Inspect current builder instance
```

**Comment by aaoafk** [comment]
for what its worth i created a `cloud.rb` file in this repo https://github.com/aaoafk/kamal-cloud-patch that works with a `kamal build push` using `driver: cloud` it is probably more complicated than it needs to be which is why im hesitant to open a PR until i simplify it somewhat but it might work for you.

this would be a nice feature to have especially in low-bandwidth environments. i was in the Dominican Republic and the internet speed was about 5 MB/s and was getting TLS failures related to byte size, i believe that was probably just due to the amount of files being sent over the wire when building an image

**Comment by matthewbjones** [comment]
Thanks for this discussion. I too was looking for an easy way to plug in Docker Build Cloud, as I'm running on Apple Silicon but targeting amd64 (droplets on Digital Ocean), and my Dockerfile needs to build ImageMagick from source (for reasons), and this can take nearly 30 minutes with the Docker virtualization, but is now much faster with Docker Build Cloud.

I've created a minimal PR here: https://github.com/basecamp/kamal/pull/1354
